source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,197 @@
+    auto extension = instruction.GetOperand(0).AsString();
+
+    if (required_extensions.count(extension) != 0) {
+      continue;
+    }
+
+    if (trimmed_extensions.count(extension) != 0) {
+      instructions_to_kill.push_back(&instruction);
+    }
+  }
+
+  // Killing an instruction invalidates the instruction iterator. We must gather
+  // & kill in 2 steps.
+  for (auto instruction : instructions_to_kill) {
+    context()->KillInst(instruction);
+  }
+}
+
+Pass::Status TrimCapabilitiesPass::FilterModuleCapabilities(
+    const CapabilitySet& required_capabilities) {
+  // Gather all the module capabilities and extension instructions.
+  std::vector<Instruction*> instructions_to_kill;
+  std::vector<spv::Capability> capabilities_to_trim;
+  for (auto& instruction : context()->capabilities()) {
+    spv::Capability capability =
+        static_cast<spv::Capability>(instruction.GetSingleWordOperand(0));
+    // Forbidden capability completely prevents trimming. Early exit.
+    if (kFordiddenCapabilities.count(capability) != 0) {
+      return Status::SuccessWithoutChange;
+    }
+
+    // Some capabilities cannot be safely removed. Leaving them untouched.
+    if (kUntouchableCapabilities.count(capability) != 0) {
+      continue;
+    }
+
+    // If the capability is unsupported, don't trim it.
+    if (kSupportedCapabilities.count(capability) == 0) {
+      continue;
+    }
+
+    if (required_capabilities.count(capability) != 0) {
+      continue;
+    }
+
+    capabilities_to_trim.push_back(capability);
+    instructions_to_kill.push_back(&instruction);
+  }
+  assert(capabilities_to_trim.size() == instructions_to_kill.size());
+
+  if (capabilities_to_trim.size() == 0) {
+    return Status::SuccessWithoutChange;
+  }
+
+  TrimExtensionsForCapabilities(required_capabilities, capabilities_to_trim);
+  for (auto instruction : instructions_to_kill) {
+    context()->KillInst(instruction);
+  }
+  return Status::SuccessWithChange;
+}
+
+Pass::Status TrimCapabilitiesPass::Process() {
+  const auto required_capabilities = DetermineRequiredCapabilities();
+#if !defined(NDEBUG)
+  // Debug only. We check the outputted required capabilities against the
+  // supported capabilities list. The supported capabilities list is useful for
+  // API users to quickly determine if they can use the pass or not. But this
+  // list has to remain up-to-date with the pass code. If we can detect a
+  // capability as required, but it's not listed, it means the list is
+  // out-of-sync. This method is not ideal, but should cover most cases.
+  {
+    CapabilitySet missing_capabilities;
+    std::copy_if(
+        required_capabilities.cbegin(), required_capabilities.cend(),
+        std::inserter(missing_capabilities, missing_capabilities.begin()),
+        [](auto capability) {
+          return kSupportedCapabilities.count(capability) == 0;
+        });
+    assert(missing_capabilities.size() == 0 &&
+           "Outdated kSupportedCapabilities.");
```

Comment:
This would be easier to read as a simple loop:

```
for(spv::Capability c : required_capabilities) {
  assert(kSupportedCapabilities.count(capability) != 0 && "Module is using a capability that is not supported.");
}
```

Three lines vs 9, no lambdas, and no extra data structures.

Reply:
WAYYYY better, that was bad.